File: languages/csharp/02-core/maps-and-frequency-counting/example/top-k-frequency.cs
// This helper example focuses on focusing on the frequency summary without repeating the full program flow.

// This extra example extends maps-and-frequency-counting with ranked frequency output.

using System;
using System.Collections.Generic;
using System.Linq;

// Keep this helper separate so the main example can focus on the larger idea without extra noise.
class Program
{
    static void Main()
    {
        Console.Write("How many words? ");
        if (!int.TryParse(Console.ReadLine(), out int count) || count <= 0)
        {
            Console.WriteLine("Please enter a positive number.");
            return;
        }

        Dictionary<string, int> frequency = new Dictionary<string, int>();
        for (int index = 0; index < count; index++)
        {
            Console.Write($"Word {index + 1}: ");
            string word = (Console.ReadLine() ?? string.Empty).Trim();
            frequency[word] = frequency.GetValueOrDefault(word) + 1;
        }

        Console.Write("How many top words to show? ");
        if (!int.TryParse(Console.ReadLine(), out int k) || k <= 0)
        {
            Console.WriteLine("Please enter a positive value for k.");
            return;
        }

        List<KeyValuePair<string, int>> items = frequency
            .OrderByDescending(entry => entry.Value)
            .ThenBy(entry => entry.Key)
            .Take(k)
            .ToList();

        Console.WriteLine($"\nTop {items.Count} words:");
        for (int index = 0; index < items.Count; index++)
        {
            Console.WriteLine($"{index + 1}. {items[index].Key} -> {items[index].Value}");
        }
    }
}

File: languages/cpp/02-core/maps-and-frequency-counting/example/top-k-frequency.cpp
// This helper example focuses on focusing on the frequency summary without repeating the full
// program flow.

#include <algorithm>
#include <iostream>
#include <map>
#include <string>
#include <vector>

using namespace std;

struct FrequencyItem {
    string word;
    int count;
};

int main() {
    int n = 0;
    cout << "How many words? ";
    cin >> n;

    if (n <= 0) {
        cout << "Please enter a positive number.\n";
        return 0;
    }

    map<string, int> frequency;
    for (int i = 0; i < n; ++i) {
        string word;
        cout << "Word " << (i + 1) << ": ";
        cin >> word;
        ++frequency[word];
    }

    int k = 0;
    cout << "How many top words to show? ";
    cin >> k;

    if (k <= 0) {
        cout << "Please enter a positive value for k.\n";
        return 0;
    }

    vector<FrequencyItem> items;
    items.reserve(frequency.size());
    for (const pair<const string, int>& entry : frequency) {
        items.push_back({entry.first, entry.second});
    }

    sort(items.begin(), items.end(), [](const FrequencyItem& a, const FrequencyItem& b) {
        if (a.count != b.count) {
            return a.count > b.count;
        }
        return a.word < b.word;
    });

    if (k > static_cast<int>(items.size())) {
        k = static_cast<int>(items.size());
    }

    cout << "\nTop " << k << " words:\n";
    for (int i = 0; i < k; ++i) {
        cout << (i + 1) << ". " << items[i].word << " -> " << items[i].count << '\n';
    }

    return 0;
}
